### DISCUSSION TITLE
Integrate with Obsidian Tasks plugin

### DISCUSSION THREAD

**Discussion by davish** [davish/obsidian-full-calendar#376] [Feature Requests]
Add a per-calendar setting for inline tasks in daily notes to use the format supported by [Obsidian Tasks](https://github.com/obsidian-tasks-group/obsidian-tasks) for parsing and serialization. This will allow Tasks to show up on Full Calendar and for Full Calendar tasks to show up in Task queries.

**Comment by davish** [comment]
I'm guessing this will be a popular feature request, so I am very open to feedback here :)

There are a bunch of [dates](https://obsidian-tasks-group.github.io/obsidian-tasks/getting-started/dates/) that can appear in a task. Which one should be considered the date for the task on full calendar?
- Due date
- Scheduled date
- Start date
- Daily Note date

I don't think Start date is relevant for the purposes of calendar placement, since it doesn't correspond to when a task is actually going to be completed.

Priority of remaining dates:
1. ⏳ Scheduled date
2. 📅 Due date if ⏳ is not defined
3. Daily note date if 📅 is not defined

**Comment by davish** [comment]
The Tasks plugin doesn't have any syntax for specifying the time of an event, so the `startTime`/`endTime`/`allDay` parameters will remain in use.

**Reply by jreus** [reply]
What's wrong with the format [Day Planner](https://github.com/ivan-lednev/obsidian-day-planner) uses? Day Planner uses ⏳ Scheduled date to determine the calendar date of the task, and prepends each task with a time range in the form of `10:00 - 13:00` 

So an example task that Daily Planner could read would look like this:

```
- [ ] 10:00 - 13:00 My Time Blocked Task ⏳ 2024-02-20
```
If a time is NOT specified, you can assume that it's an all-day calendar item. If there is no scheduled date you ignore it. This has the added benefit of producing very human-readable plaintext task lists in your daily notes. It's easy to type and less cluttered with dataview parameters.

**Reply by HochuSirok** [reply]
I agree with jreus. Day Planner plugin format would be very convenient. Or you could specify time in dataview metadata format too `[time:: 13:00 - 19:00]`, `[duration: 2024-02-20 15:00 - 2024-02-22 19:00]`.

**Reply by mattie152** [reply]
I agree with this! This is the main reason why I currently use Day Planner instead of the Daily note calendar in Full Calendar.
But a property must be used, consider having it recognise a property between brackets? I tried to see if Day Planner would recognize (startTime:: HH:mm) - (endTime:: HH:mm) since that would make it compatible with Full Calendar, but unfortunately not. It would be really nice if different task management plugins had more compatible annotation systems.

As a sidenote, it would be nice in any case if Full Calendar recognized that format and if you could customize the standard format for new Daily note calendar entries. (E.g. to (startDate:: {HH:mm}) - (endDate:: {HH:mm}) {entryname}).

**Comment by davish** [comment]
How should tasks move between notes when they are moved in Full Calendar? Right now, inline events/tasks always live in the daily note corresponding to when they take place.

- If ⏳or 📅 is defined, then don't move the task to a new note, even if the date changes.
- Otherwise, move the task.

I believe this matches how the Tasks plugin works -- the file name only matters when the [default date setting](https://obsidian-tasks-group.github.io/obsidian-tasks/getting-started/use-filename-as-default-date/) is enabled.

**Comment by Zebrazilla** [comment]
> The Tasks plugin doesn't have any syntax for specifying the time of an event, so the `startTime`/`endTime`/`allDay` parameters will remain in use.

I realize mine might be a special use case of sorts, but I divide my day planner in daily notes into the sub-headings 'morning' / 'rest of the day' / 'evening'. While it complicates things in Full Calendar, something I'd like to see would be the ability to do something like this, or a variation of:
- Specify sub-headings in options that different 'startTime's would fall under, i.e. 'startTime' of between 0500 and 0900 is put under 'morning' sub-heading, 0900 to 1900 is put under 'rest of the day', 1900 to 0500 is put under 'evening'
- Directly specify a sub-heading when creating a task event, picked from sub-headings available under the heading that's specified in options as "Under heading x in daily notes"

Some questions I have regarding this that I didn't find an answer to in your ideas above (apologies if I missed it); what is the plan for detecting tasks in a daily note? Will it detect only tasks under a specific heading, similar to how adding an event puts it "Under heading", or will it be able to discern all Tasks tasks in the note?

**Comment by davish** [comment]
> Some questions I have regarding this that I didn't find an answer to in your ideas above (apologies if I missed it); what is the plan for detecting tasks in a daily note? Will it detect only tasks under a specific heading, similar to how adding an event puts it "Under heading", or will it be able to discern all Tasks tasks in the note?

@Zebrazilla, it will continue to detect tasks under a specific heading.

**Comment by jebeaudoin** [comment]
I look forward to this integration. As a journalist, the due date is the highest priority for me (I hardly ever use scheduled although I understand its value to others. It would be easy enough to create "deadline" events on my calendar and switching to using "scheduled" dates on my tasks -- and might even improve my workflow. But would it be possible to parse more than one of these task dates (since a task is not = markdown note)?

**Comment by claremacrae** [comment]
As the maintainer of Tasks, I am very happy to pair with someone who knows this plugin’s code well and has merge permissions, to make this happen. 

I can be contacted via the email in my GitHub profile.

**Comment by claremacrae** [comment]
If there was a use case for it, I would be happy to make tasks to offer the choice of which type(s) of dates should be inferred from file names, instead of just scheduled.

**Comment by claremacrae** [comment]
> There are a bunch of [dates](https://obsidian-tasks-group.github.io/obsidian-tasks/getting-started/dates/) that can appear in a task. Which one should be considered the date for the task on full calendar?
> 
> * Due date
> * Scheduled date
> * Start date
> * Daily Note date
> 
> I don't think Start date is relevant for the purposes of calendar placement, since it doesn't correspond to when a task is actually going to be completed.

From doing Tasks' user support, I don't think there is a single agreed answer that everyone would use to select between the available dates. So I think that might suggest allowing the user to select which one they wanted to schedule the task for.

Alternatively, here's a reply I found useful when asking a similar question in Tasks:
https://github.com/obsidian-tasks-group/obsidian-tasks/discussions/616#discussioncomment-2668002
"use the earliest of the available dates"
In other words, display a task on the earliest of its dates.

**Comment by claremacrae** [comment]
Having said the above, I do think that the order you propose in https://github.com/davish/obsidian-full-calendar/issues/219#issuecomment-1339929513 makes a lot of sense.

**Comment by davish** [comment]
> As the maintainer of Tasks, I am very happy to pair with someone who knows this plugin’s code well and has merge permissions, to make this happen.

@claremacrae, thank you for the offer! I likely won't have time to start working on this until after the New Year, but will follow up with you in January.

**Comment by davish** [comment]
The relevant feature request I submitted to the Tasks plugin: https://github.com/obsidian-tasks-group/obsidian-tasks/issues/1389

**Comment by pcause** [comment]
> There are a bunch of [dates](https://obsidian-tasks-group.github.io/obsidian-tasks/getting-started/dates/) that can appear in a task. Which one should be considered the date for the task on full calendar?
> 
>     * Due date
> 
>     * Scheduled date
> 
>     * Start date
> 
>     * Daily Note date
> I don't think Start date is relevant for the purposes of calendar placement, since it doesn't correspond to when a task is actually going to be completed.

If there is a start plus due date than I'd treat the item as occurring on each of the days.  If only one of the two show on that day.

> Priority of remaining dates:
> 
>     1. ⏳ Scheduled date
> 
>     2. 📅 Due date if ⏳ is not defined
> 
>     3. Daily note date if 📅 is not defined

Since time isn't supported you can scan the item for a regex match for: \d\d:\d\d or d:\d\d-\d:\d for a time

**Comment by pcause** [comment]
as a simple first step, can you add the due date when you create a task item.  Right now when I use the popup to add an item as a task, it creates the task on the daily note page for the specified date but the item doesn't have a due date of that day.  If you added that we'd get some integration because you imply that the specified date is the due date by putting it on the daily notes page for the date.

Today you create an item with text AAA as on 2023-03-01, then on the daily notes page for 2023-03-01 you add:

- [ ] [startTime: 08:00] [endTime:: 09:00] AAA

I'm suggesting you add the time as:

- [ ] [startTime: 08:00] [endTime:: 09:00] AAA 📅2023-03-01

I can then use tasks queries that use due dates and the calendar item created will get included.

**Comment by claremacrae** [comment]
I don't know this plugin, but in case it helps, Tasks has the ability to treat dates in file-names as Scheduled dates, if you turn that on:
https://obsidian-tasks-group.github.io/obsidian-tasks/getting-started/use-filename-as-default-date/

**Comment by claremacrae** [comment]
So if you could express your Tasks queries using scheduled - possibly with boolean OR to also pick up due - than might you be able to do what you wish with the current code?

**Comment by pcause** [comment]
thanks clare, and I knew that.  The author of the plugin asks that requests be limited to a single request per issue.  I made another request in #272 for a single note to be able to have multiple calendar items and this would be needed to pair with that request, even though it also stands on its own.

**Comment by pcause** [comment]
> So if you could express your Tasks queries using scheduled - possibly with boolean OR to also pick up due - than might you be able to do what you wish with the current code?

I already use happens in my tasks queries, and this picks up scheduled.  if full-calendar added a scheduled icon/date when adding the item that would work for me.  scheduled without times would be all day, scheduled with times is an event at that time start/end

**Comment by mkdjr** [comment]
Wanted to chime in and make a request and also thank you for the plugin! 

Tasks integration with Full Calendar would be an amazing addition to my workflow. With the recent addition of statuses, especially, seeing that represented in calendar view would be fantastic.

**Comment by wc8** [comment]
> I'm guessing this will be a popular feature request, so I am very open to feedback here :)
> 
> There are a bunch of [dates](https://obsidian-tasks-group.github.io/obsidian-tasks/getting-started/dates/) that can appear in a task. Which one should be considered the date for the task on full calendar?
> 
>     * Due date
> 
>     * Scheduled date
> 
>     * Start date
> 
>     * Daily Note date
> 
> 
> I don't think Start date is relevant for the purposes of calendar placement, since it doesn't correspond to when a task is actually going to be completed.
> 
> Priority of remaining dates:
> 
>     1. ⏳ Scheduled date
> 
>     2. 📅 Due date if ⏳ is not defined
> 
>     3. Daily note date if 📅 is not defined

Showing a task **both** scheduled **and** due date (if applicable) showing the on the calendar, daily note if they are not defined. 
Further, an option for recurring if a task has until the due date seems like a good idea, either from the start date or the scheduled date. Are repeat dates under consideration? Maybe that would be another factor for recurring.

Example scenarios:
 * A bill is due on the end of the month (📅), one aims to pay it five days early (⏳). Both days (📅,⏳) show the task.
 * Write chapter is planned for today (⏳) or every Friday (🔁) 
 * Project X is due by (📅) next month, work cannot start until Tuesday, after which every weekday (🔁), some work on it is ideal, until it is finished. 

For extra credit ... showing the date marker(s) (⏳,📅), or other visual cue for it (such as by adjusted color opacity) could be helpful

**Reply by claremacrae** [reply]
Just to let you know that there is now a Tasks API and it’s first function is to allow the Tasks Edit task modal to be used by other plugins, QuickAdd users and similar. 

And someone is now adding support to Tasks for multiple formats. 

So it’s conceivable that those two things could be tied together and the parsing code shared via an API.

**Comment by nocean** [comment]
Was there any further movement on this? Would love to see these two plugins integrated.

**Reply by bepolymathe** [reply]
Yes me too !

**Comment by claremacrae** [comment]
> The author of the plugin asks that requests be limited to a single request per issue

Looking at the top of this ticket, it started off as a single very specific thing. 

And has accumulated quite mass of other requests - and it seems, also, quite a lot of expectations too.

My experience is that the broader and vaguer a request becomes, the less likely it is to ever get implemented.

**Reply by bepolymathe** [reply]
Hi, 
You're right. As far as I'm concerned, I'm still attached to the first one. 
> This will allow Tasks to show up on Full Calendar and for Full Calendar tasks to show up in Task queries.

Seeing the tasks you've accumulated in the different notes in the calendar would be a huge plus.

**Comment by pcause** [comment]
The original is enough (darvish's first 3 posts).  Agree with claire - limit to one thing,   Let's end the discussions and let the dev get to work and know that his efforts are appreciated.  PERIOD!

**Comment by Septikai** [comment]
The ability to have tasks appear on the calendar by due date, ignoring everything else in this discussion, would be an amazing addition by itself.

**Comment by mkirkland4874** [comment]
Could the simplest first approach be to be able to select "Tasks Plugin" as a read-only Calendar option (along side Full note, Daily Note, iCloud, CalDAV, etc) and then a dialog appears where we input the task query (https://publish.obsidian.md/tasks/Queries/About+Queries) we want that calendar to display? For example, 

``` tasks
not done
priority is high
```

It would not be an integration/sync, but allow users to show their task plugin tasks in Calendar on a read-only basis based on a custom task query. If a user clicks on the task in the calendar, it takes you to the note with that task. Might be a good first step without having to go through the logic to integrate the two plugins. 

I assume it wouldn't be too difficult to query that list within the calendar plugin? @claremacrae

**Reply by claremacrae** [reply]
Nice idea.

> I assume it wouldn't be too difficult to query that list within the calendar plugin?

Interestingly, I cannot see a feature request in Tasks to augment its API to allow other plugins to run Tasks searches. I know it's been talked about, but nobody has written it up.

Feel free to write the request up, @mkirkland4874...

**Reply by mkirkland4874** [reply]
Here you go!

https://github.com/obsidian-tasks-group/obsidian-tasks/issues/2459

**Comment by pcause** [comment]
Nice idea.  Tasks does have an API to bring up its entry/edit dialogue.  So when you click, bring up the tasks dialogue using the API

**Reply by mkirkland4874** [reply]
I was thinking more of a configuration dialog in Full Calendar similar to the dialog that shows up when you select "Remote (.ics format)" and you are prompted to choose a Color and URL. In this case, instead of URL, it's a text box where you input the task query you want that calendar to display. If left blank, it would query all tasks tracked by the tasks plugin. Then all those tasks from that query would show up as All day events in Calendar.

This would allow you to add multiple calendars based on your desired queries. For example, create let's say a Priority calendar that shows all high priority tasks in red (query `priority is high; not done`), and another calendar that displays non-priority tasks in blue (query `priority is not high; not done`).

When you click on the calendar item/task, it would either take you to the task in its note page or potentially show the Edit Task dialog box that is native in tasks (that shows up when you issue the `Tasks: Create or edit` command). I'm thinking may be better to just link to the note page that has the task and keep the interface in Calendar strictly read-only. It may get confusing when you have different dialogs showing up when you click an all day event (i.e., a dialog for 'native' calendar all day tasks and a separate dialog for tasks from the 'task' plugin). Maybe the best option would be to show a dialog that tells the user this task is managed by the Tasks plugin with a link to the related note you can click.

**Comment by bepolymathe** [comment]
Hello everyone, 

 I just wanted to point out the existence of the plugin [obsidian-to-ical ](https://github.com/andrewbrereton/obsidian-to-ical-plugin) which solves this problem of integration into calendars (read-only ).

**Comment by Happypig375** [comment]
Hello everyone, @YouFoundJK's fork implemented this - https://youfoundjk.github.io/plugin-full-calendar/calendars/tasks-plugin-integration/ Great work!

**Reply by claremacrae** [reply]
> Hello everyone, @YouFoundJK's fork implemented this - https://youfoundjk.github.io/plugin-full-calendar/calendars/tasks-plugin-integration/ Great work!

@YouFoundJK - thanks for supporting Tasks format - in case you haven't seen, the images are broken in [that docs page](https://youfoundjk.github.io/plugin-full-calendar/calendars/tasks-plugin-integration/):

<img width="450"  alt="image" src="https://github.com/user-attachments/assets/ff4901c9-39ee-440d-b6e7-5daa8c89b855" />

**Reply by claremacrae** [reply]
> ![Task Backlog](../assets/task-backlog.gif)

I think you are going to want to use absolute URLs, so that views outside of GitHub.com work...

**Reply by YouFoundJK** [reply]
@claremacrae Hey thanks for the heads up. Its because I never added the gifs, I just put it as placeholder for now, will add it eventually.

**Comment by eiger3970** [comment]
I used to use tasks, however evolved to Kanban cards, which offer more functionality and ease of drag/dropping from lists To do, Doing and Done. Priority might be compatibililty with Kanban boards rather than Task plugins.
